content/advent-2016/peachpy.md

```diff
@@ -0,0 +1,377 @@
++++
+linktitle = "PeachPy"
+date = "2016-12-02T11:49:46+01:00"
+author = [ "Damian Gryski" ]
+title = "Writing Go assembly functions with PeachPy"
+series = ["Advent 2016"]
+draft = true
++++
+
+**What is PeachPy**
+
+[PeachPy](https://github.com/Maratyszcza/PeachPy) is a Python-based framework
+for writing modules in assembly.  It automates away some of the details and
+allows you to use Python to generate repetitive assembly code sequences.
+
+PeachPy supports writing modules that you can use directly from Go for x86-64.
+(It also supports NaCl and syso modules, but I won't cover those in this post.)
+
+This post is going to be mostly about what you need to know about integrating
+PeachPy and less a tutorial about PeachPy specifically.
+
+All the code for this post is at
+[github.com/dgryski/peachpy-examples](https://github.com/dgryski/peachpy-examples).
+
+**A Simple Example**
+
+Lets start with a simple function: one that takes two integers and returns their sum.
+
+[embedmd]:# (args/add.py)
+```py
+import peachpy.x86_64
+
+f1 = Argument(uint64_t)
+f2 = Argument(uint16_t)
+
+with Function("add", (f1, f2), uint64_t, target=uarch.default) as function:
+    reg_f1 = GeneralPurposeRegister64()
+    reg_f2 = GeneralPurposeRegister64()
+
+    LOAD.ARGUMENT(reg_f1, f1)
+    LOAD.ARGUMENT(reg_f2, f2)
+
+    ADD(reg_f1, reg_f2)
+
+    RETURN(reg_f1)
+```
+
+Next, we'll add a function stub with a `go:generate` directive:
+
+[embedmd]:# (args/add_stub.go)
+```go
+// +build amd64
+
+package main
+
+//go:generate python -m peachpy.x86_64 add.py -S -o add_amd64.s -mabi=goasm
+func add(a uint64, b uint16) uint64
+```
+
+When we run `go generate`, we get the following assembly code:
+
+[embedmd]:# (args/add_amd64.s asm)
+```asm
+// Generated by PeachPy 0.2.0 from add.py
+
+
+// func add(f1 uint64, f2 uint16) uint64
+TEXT ·add(SB),4,$0-24
+	MOVQ f1+0(FP), AX
+	MOVWQZX f2+8(FP), BX
+	ADDQ BX, AX
+	MOVQ AX, ret+16(FP)
+	RET
+```
+
+Let's go back through the PeachPy code piece-by-piece.
+
+[embedmd]:# (args/add.py /f1/ /(?m:^$)/)
+```py
+f1 = Argument(uint64_t)
+f2 = Argument(uint16_t)
+```
+
+First, we declare our two arguments.  The types we declare here are used to
+determine where and how the arguments are passed to the function, and for the
+auto-generated function prototype.  The Go the function prototype is only in a
+comment in the generated assembly file.  There is no capability to match these
+arguments to the actual function stub, so a mismatch here won't be detected.
+
+[embedmd]:# (args/add.py /with.*/)
+```py
+with Function("add", (f1, f2), uint64_t, target=uarch.default) as function:
```

Comment:
If the target instruction set is `uarch.default`, you can omit the `target` parameter